mem/ctrl/ContextSwitchRTL.py

```diff
@@ -0,0 +1,95 @@
+"""
+==========================================================================
+ContextSwitchRTL.py
+==========================================================================
+Records/resumes progress (itertion) for functional unit (FU).
+
+Author : Yufei Yang
+  Date : Aug 11, 2025
+"""
+from pymtl3.stdlib.primitive import *
+from ...lib.basic.val_rdy.ifcs import ValRdyRecvIfcRTL as RecvIfcRTL
+from ...lib.basic.val_rdy.ifcs import ValRdySendIfcRTL as SendIfcRTL
+from ...lib.cmd_type import *
+from ...lib.opt_type import *
+from ...lib.util.common import *
+
+class ContextSwitchRTL(Component):
+
+  def construct(s, CgraPayloadType, DataType, CtrlType, num_status = 3):
+
+    # Constant
+    StatusType = mk_bits(clog2(num_status))
+
+    # Interface
+    s.recv_cmd = RecvIfcRTL(CgraPayloadType)
+    s.recv_opt = RecvIfcRTL(CtrlType)
+    s.progress_in = RecvIfcRTL(DataType)
+    s.progress_out = SendIfcRTL(DataType)
+   
+    # Component
+    s.progress_reg = Reg(DataType)
+    s.status_reg = Reg(StatusType)
+    #s.condition = Wire(Bits3)
+    # How to create a wire that can be bit sliced?
+    s.condition_2 = Wire(b1)
+    s.condition_1 = Wire(b1)
+    s.condition_0 = Wire(b1)
+
+    @update
+    def update_msg():
+      # Update condition.
+      # condition = 3'b111 (3'd7): Records the progress.
+      # condition = 3'b001 (3'd1): Resumes the progress.
+      s.condition_2 @= (s.progress_reg.out == DataType(0))
+      s.condition_1 @= (s.status_reg.out == StatusType(3))
+      s.condition_0 @= s.recv_opt.val & (s.recv_opt.msg.operation == OPT_PHI_CONST)
+
+      # Updates recv_opt.
+      s.recv_opt.rdy @= 1
+
+      # Updates recv_cmd
+      s.recv_cmd.rdy @= 1
+
+      # Updates the status register.
+      if (s.recv_cmd.val & (s.recv_cmd.msg.cmd == CMD_PAUSE)):
+        # PAUSING: 2'b11
+        s.status_reg.in_ @= StatusType(3)
+      elif (s.recv_cmd.val & (s.recv_cmd.msg.cmd == CMD_RESUME)):
+        # RESUMING: 2'b10
+        s.status_reg.in_ @= StatusType(2)
+      elif (s.recv_cmd.val & (s.recv_cmd.msg.cmd == CMD_LAUNCH)):
+        # RUNNING: 2'b00
+        s.status_reg.in_ @= StatusType(0)
+      else:
+        # Keeps unchanged.
+        s.status_reg.in_ @= s.status_reg.out
+
+      # Updates the progress register.
+      s.progress_in.rdy @= 1
+      if (s.condition_2 & s.condition_1 & s.condition_0 & s.progress_in.val):
+        # Records the progress.
+        s.progress_reg.in_ @= s.progress_in.msg
```

Comment:
I believe `progress_in` will be connected to: https://github.com/tancheng/VectorCGRA/blob/96b8adfa46a51b74c6913497e141b667d5c4bf08/mem/ctrl/CtrlMemDynamicRTL.py#L55

 - Am I right? I am confused why you declare it with `DataType` though.
 - And do you think only `progress` itself is enough to resume a paused task, do we need all the other status in the ctrl mem? (open question, no need to complete in current PR): https://github.com/tancheng/VectorCGRA/blob/96b8adfa46a51b74c6913497e141b667d5c4bf08/mem/ctrl/CtrlMemDynamicRTL.py#L62-L66

Reply:
I can't recall that, do we need both iterator and the phi_const's value? It is not clear in your PR's description, can we try to sync again on this? It has been too long. Plz provide ETA, if you are not able to make it, plz let me know.